File: ios/CamberRedline/CamberRedline/Models/TruthGraph.swift
import Foundation

struct TruthGraphHydration: Decodable {
    let callsRaw: Bool
    let interactions: Bool
    let conversationSpans: Bool
    let evidenceEvents: Bool
    let spanAttributions: Bool
    let journalClaims: Bool
    let reviewQueue: Bool

    enum CodingKeys: String, CodingKey {
        case callsRaw = "calls_raw"
        case interactions
        case conversationSpans = "conversation_spans"
        case evidenceEvents = "evidence_events"
        case spanAttributions = "span_attributions"
        case journalClaims = "journal_claims"
        case reviewQueue = "review_queue"
    }
}

struct TruthGraphSuggestedRepair: Decodable, Identifiable, Hashable {
    let action: String
    let label: String
    let idempotencyKey: String

    var id: String { idempotencyKey }

    enum CodingKeys: String, CodingKey {
        case action
        case label
        case idempotencyKey = "idempotency_key"
    }
}

struct TruthGraphResponse: Decodable {
    let ok: Bool
    let interactionId: String
    let hydration: TruthGraphHydration
    let lane: String
    let suggestedRepairs: [TruthGraphSuggestedRepair]
    let warnings: [String]

    enum CodingKeys: String, CodingKey {
        case ok
        case interactionId = "interaction_id"
        case hydration
        case lane
        case suggestedRepairs = "suggested_repairs"
        case warnings
    }
}

struct TruthGraphRepairResponse: Decodable {
    let ok: Bool
    let requestId: String?
    let status: String?
    let idempotentReplay: Bool?
    let errorCode: String?
    let error: String?

    enum CodingKeys: String, CodingKey {
        case ok
        case requestId = "request_id"
        case status
        case idempotentReplay = "idempotent_replay"
        case errorCode = "error_code"
        case error
    }
}

File: supabase/functions/redline-thread/truth_graph.ts
export type TruthGraphLane =
  | "process-call"
  | "segment-call"
  | "ai-router"
  | "journal"
  | "sms-ingest"
  | "unknown";

export type TruthGraphRepairAction =
  | "repair_process_call"
  | "repair_ai_router";

export interface TruthGraphHydration {
  calls_raw: boolean;
  interactions: boolean;
  conversation_spans: boolean;
  evidence_events: boolean;
  span_attributions: boolean;
  journal_claims: boolean;
  review_queue: boolean;
}

export interface TruthGraphSuggestedRepair {
  action: TruthGraphRepairAction;
  label: string;
  idempotency_key: string;
}

export interface TruthGraphComputationResult {
  lane: TruthGraphLane;
  suggested_repairs: TruthGraphSuggestedRepair[];
  warnings: string[];
}

export function buildRepairIdempotencyKey(
  interactionId: string,
  repairAction: TruthGraphRepairAction,
  now: Date = new Date(),
): string {
  const day = now.toISOString().slice(0, 10); // YYYY-MM-DD (UTC)
  return `repair:${repairAction}:${interactionId}:${day}`;
}

export function computeTruthGraph(
  interactionId: string,
  hydration: TruthGraphHydration,
  opts: { interaction_channel?: string | null; now?: Date } = {},
): TruthGraphComputationResult {
  const warnings: string[] = [];
  const now = opts.now ?? new Date();
  const channel = String(opts.interaction_channel || "").trim().toLowerCase();
  const looksLikeSms = interactionId.startsWith("sms_thread_") || interactionId.startsWith("sms_thread__") ||
    channel === "sms" || channel === "text" || channel === "sms_thread";

  let lane: TruthGraphLane = "unknown";

  if (looksLikeSms) {
    lane = "sms-ingest";
  } else if (!hydration.calls_raw || !hydration.interactions) {
    lane = "process-call";
  } else if (!hydration.conversation_spans || !hydration.evidence_events) {
    lane = "segment-call";
  } else if (!hydration.span_attributions) {
    lane = "ai-router";
  } else if (!hydration.journal_claims) {
    lane = "journal";
  } else {
    lane = "unknown";
  }

  const suggested_repairs: TruthGraphSuggestedRepair[] = [];

  if (lane === "process-call" || lane === "segment-call") {
    suggested_repairs.push({
      action: "repair_process_call",
      label: "Replay process-call (rebuild raw → interaction → spans)",
      idempotency_key: buildRepairIdempotencyKey(interactionId, "repair_process_call", now),
    });
    if (!hydration.calls_raw) {
      warnings.push("missing_calls_raw_snapshot: repair_process_call may be unable to reconstruct payload");
    }
  }

  if (lane === "ai-router") {
    suggested_repairs.push({
      action: "repair_ai_router",
      label: "Re-run attribution/router (reseed spans → reroute)",
      idempotency_key: buildRepairIdempotencyKey(interactionId, "repair_ai_router", now),
    });
  }

  if (lane === "journal") {
    warnings.push("journal_missing: no v0 repair hook exposed (manual follow-up)");
  }

  return { lane, suggested_repairs, warnings };
}

